File: internal/statusbar/statusbar.go
// internal/statusbar/statusbar.go
package statusbar

import (
	"fmt"
	"sync"
)

const (
	cursorSave    = "\x1b7"
	cursorRestore = "\x1b8"
	sgrReset      = "\x1b[0m"
	fgGreen       = "\x1b[32m"
	fgRed         = "\x1b[31m"
	fgYellow      = "\x1b[33m"
)

// labelWidth is the fixed width of the status label (widest possible label).
// " [yoyo: on Xs | RuleName]" — we reserve space for "on Xs" + rule.
// Keep it wide enough; truncate rule name if needed.
const minLabelWidth = 22 // " [yoyo: on Xs | ...]  " minimum

// StatusBar renders a bottom-right ANSI overlay around PTY output frames.
// All methods are goroutine-safe (SIGWINCH resize races with event loop writes).
type StatusBar struct {
	mu        sync.Mutex
	rows      uint16
	cols      uint16
	enabled   bool
	delaySecs int
	countdown int    // remaining seconds; -1 = no active countdown
	rule      string
	painted   bool
	midSeq    bool
	prefix    bool // true while waiting for Ctrl+Y command byte
	dryRun    bool
	buf       []byte // reusable output buffer
}

// New creates a StatusBar. enabled=true means auto-approve is active.
func New(rows, cols uint16, enabled bool, delaySecs int) *StatusBar {
	return &StatusBar{
		rows:      rows,
		cols:      cols,
		enabled:   enabled,
		delaySecs: delaySecs,
		countdown: -1,
	}
}

func (sb *StatusBar) Toggle() {
	sb.mu.Lock()
	sb.enabled = !sb.enabled
	sb.mu.Unlock()
}

func (sb *StatusBar) SetDelay(secs int) {
	sb.mu.Lock()
	sb.delaySecs = secs
	sb.mu.Unlock()
}

// SetRule sets the last-matched rule name shown in the label.
func (sb *StatusBar) SetRule(rule string) {
	sb.mu.Lock()
	sb.rule = rule
	sb.mu.Unlock()
}

// SetCountdown sets the remaining seconds shown during an active approval timer.
// Pass -1 to clear the countdown display.
func (sb *StatusBar) SetCountdown(secs int) {
	sb.mu.Lock()
	sb.countdown = secs
	sb.mu.Unlock()
}

// SetPrefix sets the Ctrl+Y prefix-waiting indicator.
func (sb *StatusBar) SetPrefix(active bool) {
	sb.mu.Lock()
	sb.prefix = active
	sb.mu.Unlock()
}

// SetDryRun enables dry-run mode display.
func (sb *StatusBar) SetDryRun(on bool) {
	sb.mu.Lock()
	sb.dryRun = on
	sb.mu.Unlock()
}

// Resize updates terminal dimensions.
func (sb *StatusBar) Resize(rows, cols uint16) {
	sb.mu.Lock()
	sb.rows = rows
	sb.cols = cols
	sb.mu.Unlock()
}

// WrapFrame injects clear-previous and paint-new ANSI sequences around frame.
// Returns a single buffer for one atomic write to stdout.
func (sb *StatusBar) WrapFrame(frame []byte) []byte {
	sb.mu.Lock()
	defer sb.mu.Unlock()
	prevMid := sb.midSeq
	sb.midSeq = endsMidEscape(frame) || endsMidUTF8(frame)

	label := sb.labelText()
	lw := uint16(len(label))
	if sb.cols < lw+2 || sb.rows == 0 {
		return frame // terminal too narrow
	}

	col := sb.cols - lw

	var clear []byte
	if sb.painted && !prevMid {
		sb.painted = false
		blank := make([]byte, lw)
		for i := range blank {
			blank[i] = ' '
		}
		clear = overlayAt(sb.rows, col, "", string(blank))
	} else if prevMid {
		sb.painted = false
	}

	var paint []byte
	if !sb.midSeq {
		sb.painted = true
		color := sb.labelColor()
		paint = overlayAt(sb.rows, col, color, label)
	}

	// Reuse buffer to reduce allocations
	need := len(clear) + len(frame) + len(paint)
	if cap(sb.buf) < need {
		sb.buf = make([]byte, 0, need+256)
	}
	sb.buf = sb.buf[:0]
	sb.buf = append(sb.buf, clear...)
	sb.buf = append(sb.buf, frame...)
	sb.buf = append(sb.buf, paint...)
	return sb.buf
}

func (sb *StatusBar) labelColor() string {
	if sb.prefix {
		return fgYellow
	}
	if !sb.enabled {
		return fgRed
	}
	if sb.dryRun {
		return fgYellow
	}
	if sb.countdown >= 0 {
		return fgYellow // counting down — attention
	}
	return fgGreen
}

func (sb *StatusBar) labelText() string {
	if sb.prefix {
		return " [yoyo: ^Y …] "
	}
	if !sb.enabled {
		if sb.dryRun {
			return " [yoyo: dry off] "
		}
		return " [yoyo: off] "
	}
	mode := "on"
	if sb.dryRun {
		mode = "dry"
	}
	// Active countdown: show remaining seconds
	if sb.countdown >= 0 {
		if sb.rule == "" {
			return fmt.Sprintf(" [yoyo: %s %ds] ", mode, sb.countdown)
		}
		return fmt.Sprintf(" [yoyo: %s %ds | %s] ", mode, sb.countdown, sb.rule)
	}
	if sb.rule == "" {
		return fmt.Sprintf(" [yoyo: %s %ds] ", mode, sb.delaySecs)
	}
	return fmt.Sprintf(" [yoyo: %s %ds | %s] ", mode, sb.delaySecs, sb.rule)
}

func overlayAt(row, col uint16, color, text string) []byte {
	s := fmt.Sprintf("%s\x1b[%d;%dH%s%s%s%s",
		cursorSave, row, col, sgrReset, color, text, cursorRestore)
	return []byte(s)
}

// endsMidEscape reports whether data ends with an incomplete ANSI escape.
func endsMidEscape(data []byte) bool {
	pos := -1
	for i := len(data) - 1; i >= 0; i-- {
		if data[i] == 0x1b {
			pos = i
			break
		}
	}
	if pos < 0 {
		return false
	}
	tail := data[pos:]
	if len(tail) == 1 {
		return true
	}
	switch tail[1] {
	case '[':
		for _, b := range tail[2:] {
			if b >= 0x40 && b <= 0x7E {
				return false
			}
		}
		return true
	case ']':
		for i := 0; i < len(tail)-1; i++ {
			if tail[i] == 0x07 {
				return false
			}
			if tail[i] == 0x1b && tail[i+1] == '\\' {
				return false
			}
		}
		return true
	}
	return false
}

// endsMidUTF8 reports whether data ends with an incomplete UTF-8 character.
func endsMidUTF8(data []byte) bool {
	if len(data) == 0 {
		return false
	}
	i := len(data) - 1
	for i > 0 && (data[i]&0xC0) == 0x80 {
		i--
	}
	b := data[i]
	var expected int
	switch {
	case b < 0x80:
		expected = 1
	case b < 0xE0:
		expected = 2
	case b < 0xF0:
		expected = 3
	default:
		expected = 4
	}
	return (len(data) - i) < expected
}

File: auto-approve/src/status_bar.rs
use std::fmt::Write;

const LABEL_OFF: &str = " [auto-approve: off] ";

const CURSOR_SAVE: &str = "\x1b[s";
const CURSOR_RESTORE: &str = "\x1b[u";
const SGR_RESET: &str = "\x1b[0m";
const FG_RED: &str = "\x1b[31m";
const FG_GREEN: &str = "\x1b[32m";

pub struct StatusBar {
    rows: u16,
    cols: u16,
    enabled: bool,
    delay_secs: u32,
    label_width: u16,
    /// Whether the label is currently drawn on screen.
    painted: bool,
    /// Whether the previous frame left the terminal mid-escape or mid-UTF-8.
    mid_sequence: bool,
}

impl StatusBar {
    pub fn new(rows: u16, cols: u16, enabled: bool, delay_secs: u32) -> Self {
        let mut bar = StatusBar {
            rows,
            cols,
            enabled,
            delay_secs,
            label_width: 0,
            painted: false,
            mid_sequence: false,
        };
        bar.recompute_label_width();
        bar
    }

    pub fn toggle(&mut self) {
        self.enabled = !self.enabled;
    }

    pub fn set_delay(&mut self, secs: u32) {
        self.delay_secs = secs;
    }

    pub fn resize(&mut self, rows: u16, cols: u16) {
        self.rows = rows;
        if self.cols != cols {
            self.cols = cols;
            self.recompute_label_width();
        }
    }

    /// Wrap a PTY frame with clear (prefix) and paint (suffix) sequences,
    /// returned as a single buffer for one atomic `write_all` call.
    ///
    /// Skips injection when the output stream is mid-escape or mid-UTF-8 to
    /// avoid corrupting multi-byte characters or escape sequences that span
    /// frame boundaries.
    pub fn wrap_frame(&mut self, frame: &[u8]) -> Vec<u8> {
        let prev_mid = self.mid_sequence;
        self.mid_sequence = ends_mid_escape(frame) || ends_mid_utf8(frame);

        if self.label_width == 0 || self.rows == 0 {
            return frame.to_vec();
        }

        let label_col = self.cols - self.label_width + 1;

        // Phase 1: clear previous label.
        // Skip if previous frame left us mid-sequence (current frame starts
        // with continuation bytes — injecting would corrupt them).
        let clear = if self.painted && !prev_mid {
            self.painted = false;
            let blank: String = std::iter::repeat(' ').take(self.label_width as usize).collect();
            overlay_at(self.rows, label_col, "", &blank)
        } else {
            if prev_mid {
                self.painted = false;
            }
            Vec::new()
        };

        // Phase 2: paint new label after the frame.
        // Skip if this frame ends mid-sequence.
        let paint = if self.mid_sequence {
            Vec::new()
        } else {
            self.painted = true;
            let fg_color = if self.enabled { FG_RED } else { FG_GREEN };
            overlay_at(self.rows, label_col, fg_color, &self.label_text())
        };

        let mut buf = Vec::with_capacity(clear.len() + frame.len() + paint.len());
        buf.extend_from_slice(&clear);
        buf.extend_from_slice(frame);
        buf.extend_from_slice(&paint);
        buf
    }

    fn label_text(&self) -> String {
        let label = if self.enabled {
            format!(" [auto-approve: on {}s] ", self.delay_secs)
        } else {
            LABEL_OFF.to_string()
        };
        let on_width = " [auto-approve: on 0s] ".len();
        format!("{:>width$}", label, width = on_width)
    }

    fn recompute_label_width(&mut self) {
        let on_width = " [auto-approve: on 0s] ".len() as u16;
        if self.cols < on_width {
            self.label_width = 0;
        } else {
            self.label_width = on_width;
        }
    }
}

/// Write `text` at the given `row` and `col` with optional SGR color, wrapped
/// in cursor save/restore so the cursor position is preserved.
fn overlay_at(row: u16, col: u16, color: &str, text: &str) -> Vec<u8> {
    let mut s = String::with_capacity(48 + text.len());
    s.push_str(CURSOR_SAVE);
    let _ = write!(s, "\x1b[{};{}H", row, col);
    s.push_str(SGR_RESET);
    s.push_str(color);
    s.push_str(text);
    s.push_str(SGR_RESET);
    s.push_str(CURSOR_RESTORE);
    s.into_bytes()
}

/// Check if `data` ends with an incomplete ANSI escape sequence.
fn ends_mid_escape(data: &[u8]) -> bool {
    let Some(esc_pos) = data.iter().rposition(|&b| b == 0x1b) else {
        return false;
    };
    let tail = &data[esc_pos..];

    if tail.len() == 1 {
        return true;
    }

    match tail[1] {
        // CSI: ESC [ <params> <terminator 0x40-0x7E>
        b'[' => !tail[2..].iter().any(|&b| (0x40..=0x7E).contains(&b)),
        // OSC: ESC ] ... terminated by BEL (0x07) or ST (ESC \)
        b']' => !tail[2..].iter().any(|&b| b == 0x07)
            && !data[esc_pos + 1..].windows(2).any(|w| w == b"\x1b\\"),
        _ => false,
    }
}

/// Check if `data` ends with an incomplete multi-byte UTF-8 character.
fn ends_mid_utf8(data: &[u8]) -> bool {
    if data.is_empty() {
        return false;
    }
    // Walk backwards from the end to find the start byte of the last character.
    let mut i = data.len() - 1;
    // Skip continuation bytes (10xxxxxx).
    while i > 0 && (data[i] & 0xC0) == 0x80 {
        i -= 1;
    }
    let start = data[i];
    let expected_len = if start < 0x80 {
        1
    } else if start < 0xE0 {
        2
    } else if start < 0xF0 {
        3
    } else {
        4
    };
    (data.len() - i) < expected_len
}
